### PR TITLE
Add pass to replace some ops with channels last variants.

### LINKED ISSUE #20094: [Memory-layout infra] Create ToContiguousChannelsLastPass
Part of #19299. Create a single common pass in `backends/transforms` which replaces corresponding ATen kernels with the
`exir_ops.edge.channels_last` dialect surrounded by permutes, and removes dim-order changing ops.

[AtenToDialectPAss](https://github.com/pytorch/executorch/blob/main/backends/transforms/aten_to_dialect_pass.py) may be used as a base.

### PR DESCRIPTION
**Pull request by MartinPavella**
### Summary
Add a transformation pass to replace some operators with their channels last variant. Currently only `convolution` and `avg_pool2d` are implemented, as those are the only operators with a channels last variant. In the future, the pass can be easily extended to support more operators.

### Test plan
Tested by `pytest backends/transforms/test/test_replace_ops_with_channels_last_variants.py`.

### CODE REVIEW

**Comment by MartinPavella** [comment]
@AdrianLundell this is the first part of https://github.com/pytorch/executorch/issues/20094. I am planning on introducing the second part (a pass which makes the graph contiguous) separately.

File: backends/transforms/replace_ops_with_channels_last_variants.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+from executorch.exir import ExportedProgram
+from executorch.exir.dialects._ops import ops as exir_ops
+from torch.fx.node import Target
+from torch.fx.passes.infra.pass_manager import PassResult
+
+_NCHW_TO_NHWC_PERM: list[int] = [0, 2, 3, 1]
+_NHWC_TO_NCHW_PERM: list[int] = [0, 3, 1, 2]
+
+
+def _requires_rank(rank: int) -> Callable[[torch.fx.Node], bool]:
```
Comment (AdrianLundell) [inline backends/transforms/replace_ops_with_channels_last_variants.py]: Why not allow (CHW,) -> (HWC) as well?

**Comment by AdrianLundell** [comment]
Great to see the collaboration on this, just one question!

File: backends/transforms/replace_ops_with_channels_last_variants.py (hunk 1 above)
Comment (MartinPavella) [reply backends/transforms/replace_ops_with_channels_last_variants.py]: Good point, I missed that use case. 
I will make the relevant changes.

File: backends/transforms/replace_ops_with_channels_last_variants.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from dataclasses import dataclass
+from typing import Callable
+
+import executorch.backends.transforms.channels_last_ops  # noqa: F401
+
+import torch
+from executorch.backends.xnnpack._passes.xnnpack_pass import ExportPass
```
Comment (rascani) [inline backends/transforms/replace_ops_with_channels_last_variants.py]: `from executorch.exir.pass_base import ExportPass`

File: backends/transforms/replace_ops_with_channels_last_variants.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                continue
+            contiguous_dim_order = tuple(range(len(node.meta["val"].shape)))
+            if node.meta["val"].dim_order() != contiguous_dim_order:
+                continue
+            if spec.filter_fn is not None and not spec.filter_fn(node):
+                continue
+
+            args = list(node.args)
+
+            with graph.inserting_before(node):
```
Comment (rascani) [inline backends/transforms/replace_ops_with_channels_last_variants.py]: Might want to consider using the `super().call_operator()` style of rewrites, which eliminates having to deal with meta & graph cleanup. Not a blocker though.

```
      def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
          self._modified = False
          result = super().call(graph_module)
          return PassResult(result.graph_module, self._modified)

      def call_operator(self, op, args, kwargs, meta: NodeMetadata) -> ProxyValue:
          spec = self.op_map.get(op)
          if spec is None:
              return super().call_operator(op, args, kwargs, meta)

          val = meta["val"]
          contiguous_dim_order = tuple(range(len(val.shape)))
          if val.dim_order() != contiguous_dim_order or (
              spec.filter_fn is not None and not spec.filter_fn(val)
          ):
              return super().call_operator(op, args, kwargs, meta)

          new_args = list(args)
          for idx in spec.input_indices:
              new_args[idx] = super().call_operator(
                  exir_ops.edge.channels_last.permute_copy.default,
                  (new_args[idx], _NCHW_TO_NHWC_PERM),
                  {},
                  meta,
              )

          nhwc_out = super().call_operator(spec.target, tuple(new_args), kwargs, meta)

          self._modified = True
          return super().call_operator(
              exir_ops.edge.channels_last.permute_copy.default,
              (nhwc_out, _NHWC_TO_NCHW_PERM),
              {},
              meta,
          )
```

**Review by rascani — COMMENTED** [review]
Can you add BUCK support in the targets.bzl?

```
      runtime.python_library(
          name = "replace_ops_with_channels_last_variants",
          srcs = [
              "replace_ops_with_channels_last_variants.py",
          ],
          visibility = [
              "//executorch/backends/...",
          ],
          deps = [
              "//caffe2:torch",
              ":channels_last_ops",
              "//executorch/exir:pass_base",
          ],
      )

      runtime.python_test(
          name = "test_replace_ops_with_channels_last_variants",
          srcs = [
              "test/test_replace_ops_with_channels_last_variants.py",
          ],
          deps = [
              "//caffe2:torch",
              ":channels_last_ops",
              ":replace_ops_with_channels_last_variants",
              "//executorch/exir:lib",
              "fbsource//third-party/pypi/pytest:pytest",
          ],
      )
```

File: backends/transforms/replace_ops_with_channels_last_variants.py (hunk 2 above)
Comment (MartinPavella) [reply backends/transforms/replace_ops_with_channels_last_variants.py]: Thank you. Fixed

File: backends/transforms/replace_ops_with_channels_last_variants.py (hunk 3 above)
Comment (MartinPavella) [reply backends/transforms/replace_ops_with_channels_last_variants.py]: Thank you for the suggestion. After consideration, I prefer your solution. I have updated the implementation.

File: backends/transforms/replace_ops_with_channels_last_variants.py (hunk 1 above)
Comment (MartinPavella) [reply backends/transforms/replace_ops_with_channels_last_variants.py]: For `convolution`, `unsqueeze` and `squeeze` operators are inserted during lowering to edge, which explicitly add/remove the batch size of `1`, and the exisitng code with `4D` permutations works correctly.
For `avg_pool2d`, this doesn't happen. 

I have decided to update this pass to also insert the `[un]squeeze` for implicit batch cases, which should solve all related issues while remaining consistent with `convolution`.

**Comment by MartinPavella** [comment]
Implemented changes based on reviews. 
I plan on updating this PR once more to add support for the new channels last ops recently added by R. J.: https://github.com/pytorch/executorch/pull/20775

File: backends/transforms/replace_ops_with_channels_last_variants.py (hunk 3 above)
Comment (MartinPavella) [reply backends/transforms/replace_ops_with_channels_last_variants.py]: Update: When I tried to add support for `max_pool2d_with_indices`, I realized that the `call_operator()` style won't work. This is because `permute` and `squeeze` operators must be added after the `getitem` nodes of the `max_pool2d_with_indices`, so the resulting sugbraph would have multiple distinct outputs (which as far as I can tell is incompatible with the `call_operator()` pass style).

Therefore, I have switched back to the original approach.

**Review by AdrianLundell — APPROVED** [review]
Nice, it is a bit unfortunate that there are these arbitrary differences between ops which must be handled everywhere but not much to do about it.

**Comment by rascani** [comment]
@MartinPavella  - Would you mind taking care of the EasyCLA? This is part of the migration of ExecuTorch to the PyTorch/Linux Foundation.

**Review by rascani — APPROVED** [review]

**Comment by robert-kalmar** [comment]
> @MartinPavella - Would you mind taking care of the EasyCLA? This is part of the migration of ExecuTorch to the PyTorch/Linux Foundation.

@rascani , let me handle it. Working with legal to get it signed.
